kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/config/FormatSelection.scala

```diff
@@ -0,0 +1,179 @@
+  ): FormatSelection = {
+    val withoutTicks = formatAsString.replace("`", "")
+    val split        = withoutTicks.split("_")
+
+    val formatOptions: Set[FormatOptions] = if (split.size > 1) {
+      split.splitAt(1)._2.flatMap(FormatOptions.withNameInsensitiveOption).toSet
+    } else {
+      Set.empty
+    }
+
+    val format = Format
+      .withNameInsensitiveOption(split(0))
+    format.getOrElse(throw new IllegalArgumentException(s"Unsupported format - $formatAsString")) match {
+      case Format.Json    => JsonFormatSelection
+      case Format.Avro    => AvroFormatSelection
+      case Format.Parquet => ParquetFormatSelection
+      case Format.Text    => TextFormatSelection(readTextMode())
+      case Format.Csv     => CsvFormatSelection(formatOptions)
+      case Format.Bytes   => BytesFormatSelection(formatOptions)
+    }
+  }
+
+}
+
+case object JsonFormatSelection extends FormatSelection {
+  override def toStreamReader(
+    inputStream:          InputStream,
+    fileSize:             Long,
+    bucketAndPath:        S3Location,
+    recreateInputStreamF: () => Either[Throwable, InputStream],
+  ) =
+    new TextFormatStreamReader(inputStream, bucketAndPath)
+
+  override def extension: String = "json"
+}
+
+case object AvroFormatSelection extends FormatSelection {
+  override def availableCompressionCodecs: Map[CompressionCodecName, Boolean] = Map(
+    UNCOMPRESSED -> false,
+    DEFLATE      -> true,
+    BZIP2        -> false,
+    SNAPPY       -> false,
+    XZ           -> true,
+    ZSTD         -> true,
+  )
+
+  override def toStreamReader(
+    inputStream:          InputStream,
+    fileSize:             Long,
+    bucketAndPath:        S3Location,
+    recreateInputStreamF: () => Either[Throwable, InputStream],
+  ) =
+    new AvroFormatStreamReader(inputStream, bucketAndPath)
+
+  override def extension: String = "avro"
+}
+
+case object ParquetFormatSelection extends FormatSelection {
+  override def availableCompressionCodecs: Map[CompressionCodecName, Boolean] = Set(
+    UNCOMPRESSED,
+    SNAPPY,
+    GZIP,
+    LZO,
+    BROTLI,
+    LZ4,
+    ZSTD,
+  ).map(_ -> false).toMap
+
+  override def toStreamReader(
+    inputStream:          InputStream,
+    fileSize:             Long,
+    bucketAndPath:        S3Location,
+    recreateInputStreamF: () => Either[Throwable, InputStream],
+  ) = ParquetFormatStreamReader.apply(inputStream, fileSize, bucketAndPath, recreateInputStreamF)
+
+  override def extension: String = "parquet"
+}
+case class TextFormatSelection(readTextMode: Option[ReadTextMode]) extends FormatSelection {
+  override def toStreamReader(
+    inputStream:          InputStream,
+    fileSize:             Long,
+    bucketAndPath:        S3Location,
+    recreateInputStreamF: () => Either[Throwable, InputStream],
+  ): S3FormatStreamReader[StringSourceData] =
+    readTextMode.map(rtm => rtm.createStreamReader(inputStream, bucketAndPath)).getOrElse(new TextFormatStreamReader(
```

Comment:
Would it not improve the code to have a TextFromatStreamReader to be built like the other two and avoid this code being responsible for building the readers?

Reply:
I have moved this.  Thanks!